text/2017-10-06-api-layers.md

```diff
@@ -0,0 +1,206 @@
+# Summary
+
+This RFC introduces three layers of API for supporting diversified use cases.
+
+
+# Motivation
+
+As the Crossbeam project is growing, users want more diversified API than the existing one.
+
+First, majority of users will be happy with the current API, which interacts with the default
+singleton garbage collector.
+
+Second, some users may want to create their own garbage collector, and use the collector via a
+*handle* to it. Since the current API provides only one garbage collector, Crossbeam currently does
+not support this use case.
+
+Lastly, some users want to embed a garbage collector in another systems library, e.g. memory
+allocator and thread management environment.  Note that it is often a part of the standard library,
+thus this use case requires (at least a part of) Crossbeam to be implemented in the `#[no_std]`
+environment. Likewise, Crossbeam currently does not support this use case.
+
+In order to support these diversified use cases, this author proposes to generalize the API by
+introducing an API layer for each use case: (1) **the default garbage collector API** for the first
+use case; (2) **`Collector` API** for the second one; and (3) **internal API** for the third one.
```

Comment:
Yes, you can use the `Collector` API for e.g. `elfmalloc`.  In particular, as you pointed out, `Collector` API can be easily adjusted for being used in `#[no_std]` environment.  I think the main difference between the internal and the `Collector` APIs is the number of pointer indirection.


Using the `Collector` API, `elfmalloc`'s `Global` may look like:

```rust
struct Global {
    collector: epoch::Collector,
    ... // other global data
}
```

Note that `epoch::Collector` is just `Arc<epoch::Global>` under the hood.  So there is an additional pointer indirection for accessing the global data for the underlying collector from a `elfmalloc` handle.  The situation gets worse if you want to embed `elfmalloc` in even another systems library: two pointer indirections are required for accessing the global data for the collector from the enclosing library's handle.


If you use the internal API, there is no additional indirection for accessing the global data for the collector, as evident in the following hypothetical definition of `elfmalloc`'s `Global`:

```rust
struct Global {
    epoch: epoch::Global,
    ... // other global data
}
```

As I said in the RFC, these layers of internal, `Collector`(/`Allocator`/`Logger`/...), and the default collector(/allocator/logger/...) APIs of different libraries compose nicely.


I'd like to also ask @joshlf's opinion.  He is a main developer of `elfmalloc`.

Reply:
Hmm. I agree that there is one level od indirection less, but what kind of operations would actually benefit from this?

Suppose we're pinning locals/handles.
In order to pin a `Local`, you'd do: `my_local.pin(&my_global, || ...)`.
But if using `Collector` API, then you have: `my_handle.pin(|| ...)`.
The first case has the manually supplied `&my_global`, while the second one simply reaches for the `Arc<Collector>` inside `my_handle`. Both cases are actually very similar - I'd be very surprised to see any tangible difference in performance here.

But sometimes we might avoid locals/handles and reach for the global/collector directly, for example: `my_global.collect()` or `my_global.has_garbage()`. But even then I don't expect any noticeable difference, and can hardly imagine such method calls being on hot paths in the code. :)

Reply:
I'll let @ezrosent chime in here since he's done work on [`bagpipe`](https://crates.io/crates/bagpipe) than I have (which is the code in elfmalloc that actually uses epoch gc), but I think this would work for us.

Reply:
@stjepang Thank you for the analysis.  I was handwaving on this matter and didn't think of the performance overhead precisely.  Now I know the tradeoff clearly.

Now I'm fine with removing the internal API for this RFC and the corresponding PR, provided that `elfmalloc` developers are happy with the `Collector` API.

Though I still prefer exposing the internal API for the last bit of optimization.  I think we can discuss this matter later in another RFC after we gather more real-world use cases of Crossbeam.

Reply:
Same caveat about @ezrosent knowing this better than I do, but I'm pretty sure we'd be fine with the `Collector` API so long as it can eventually support a parametric `Alloc` (which, as I mentioned in a different comment, depends on `Arc` supporting a parametric `Alloc`).

Reply:
I removed the internal API from this RFC and the corresponding PR.  I think there are no more unresolved issues, and they are ready to be merged.  Please give another round of reviews.  Thanks!

Reply:
> In order to pin a Local, you'd do: my_local.pin(&my_global, || ...).
But if using Collector API, then you have: my_handle.pin(|| ...).
The first case has the manually supplied &my_global, while the second one simply reaches for the Arc<Collector> inside my_handle.

Just to clarify, the second case reaches for `Arc<Global>` inside the handle, which is also just one level of indirection. If it reached for `Collector` first, there would be two levels.

> I removed the internal API from this RFC and the corresponding PR.

I think it's fine if we don't expose it for now and only make the `Arc`-based API available, possibly adding the internal API later.